### PR TITLE
Multi structure

### PR DESCRIPTION
**Pull request by jcass11**
Allows the mti_nma pipeline to be run on multiple structures by creating structure-named versions of every step.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,6 +2,7 @@
 /local_staging
 /data
 /dask-worker-space
+/~
```
Comment (JacksonMaxfield) [inline .gitignore]: This was caused by a bug and I _think_ I fixed it in one of my PRs to this branch. Can remove.

File: README.md [hunk 2]
```diff
@@ -160,13 +163,13 @@ The following commands are the minimal set of commands to get setup for running
 4. Run the pipeline
 
     ```bash
-    mti_nma all run --path_blender /allen/aics/modeling/jacksonb/applications/blender-2.82-linux64/blender --nsamples {int} --distributed
+    mti_nma all run --path_blender /allen/aics/modeling/jacksonb/applications/blender-2.82-linux64/blender --nsamples {int} --distributed True
```
Comment (JacksonMaxfield) [inline README.md]: You actually don't need the `True` at the end. Simply by passing `--distributed` it should assume True

File: README.md [hunk 3]
```diff
@@ -160,13 +163,13 @@ The following commands are the minimal set of commands to get setup for running
...
     ```
 
 5. (Optional) Connect to the Dask UI
 
     _In a new terminal_
 
     ```bash
-    ssh -A -J slurm-master -L {port_provided_by_log}:{node_you_are_on}:{port_provided_by_log} {node_you_are_on}
+    ssh -A -J <username>@slurm-master.corp.alleninstitute.org -L {port_provided_by_log}:{node_you_are_on}:{port_provided_by_log} <username>@{node_you_are_on}
```
Comment (JacksonMaxfield) [inline README.md]: Lol nice WFH :upside_down_face:

File: mti_nma/bin/all.py [hunk 4]
```diff
@@ -63,37 +66,50 @@ def run(
...
             exe = LocalExecutor()
             distributed_executor_address = None
+            log.info(f"Debug flagged. Will use threads instead of Dask.")
         else:
             if distributed:
                 # Create or get log dir
                 # Do not include ms
                 log_dir_name = datetime.now().isoformat().split(".")[0]
-                log_dir = Path(f"~/.dask_logs/mti_nma/{log_dir_name}")
+                log_dir = Path(f"~/.dask_logs/mti_nma/{log_dir_name}").expanduser()
```
Comment (JacksonMaxfield) [inline mti_nma/bin/all.py]: Yep this `.expanduser()` solves the `~` directory from being made. So can remove the gitignored line.

File: mti_nma/bin/all.py [hunk 5]
```diff
@@ -102,9 +118,10 @@ def run(
                     local_directory=str(log_dir),
                     log_directory=str(log_dir)
                 )
+                log.info("Created SLURMCluster")
 
-                # Scale workers
-                cluster.adapt(minimum_jobs=1, maximum_jobs=40)
+                # Set adaptive worker settings
+                cluster.scale_up(40)
```
Comment (JacksonMaxfield) [inline mti_nma/bin/all.py]: I would change this to
```suggestion
                cluster.adapt(minimum_jobs=30, maximum_jobs=60)
```

File: mti_nma/bin/all.py [hunk 6]
```diff
@@ -132,49 +151,83 @@ def run(
...
+                        **kwargs
+                    )
+
+                # If nucleus and cell membrane were anlyzed, draw comparison plot
+                if "Nuc" and "Cell" in structs:
+                    compare_nuc_cell(nma_nuc_df, nma_cell_df)
+
+            # Run flow, get ending state, and visualize pipeline
+            flow.run(executor=exe)
+            flow.visualize()
```
Comment (JacksonMaxfield) [inline mti_nma/bin/all.py]: May want to remove this line or make it a parameter `--viz`. the dependencies required for `flow.visualize()` can't just be bare installed with `pip`.

File: mti_nma/steps/compare_nuc_cell/compare_nuc_cell.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        --------------------
+        distributed_executor_address: Optional[str]
+            An optional executor address to pass to some computation engine.
+        clean: bool
+            Should the local staging directory be cleaned prior to this run.
+            Default: False (Do not clean)
+        debug: bool
+            A debug flag for the developer to use to manipulate how much data runs,
+            how it is processed, etc.
+            Default: False (Do not debug)
```
Comment (JacksonMaxfield) [inline mti_nma/steps/compare_nuc_cell/compare_nuc_cell.py]: I should have told you this so my bad, but these should be folded into the `Parameters` section of each docstring where needed. They are just always passed by default so if you don't use them then get rid of them but if you do put them in the primary `Parameters` sections.

File: mti_nma/steps/nma/nma.py [hunk 8]
```diff
@@ -90,6 +89,9 @@ def run(
             Default: None (no distributed executor, use local threads)
         """
 
+        # Set upstream task using structure-named version of previous step
+        self.direct_upstream_tasks = [Avgshape(step_name=f"avgshape_{struct}")]
```
Comment (JacksonMaxfield) [inline mti_nma/steps/nma/nma.py]: I don't think this will work...

File: mti_nma/steps/shparam/shparam.py [hunk 9]
```diff
@@ -140,6 +143,9 @@ def run(
             Default: None (no distributed executor, use local threads)
         """
 
+        # Set upstream task using structure-named version of previous step
+        self.direct_upstream_tasks = [Singlecell(step_name=f"singlecell_{struct}")]
```
Comment (JacksonMaxfield) [inline mti_nma/steps/shparam/shparam.py]: Similar here, don't think this will work.

File: setup.py [hunk 10]
```diff
@@ -64,6 +63,8 @@
     "scikit-image",
     "seaborn",
     "vtk",
+    "graphviz",
```
Comment (JacksonMaxfield) [inline setup.py]: Does this fully pip install? huh. I really thought it had to be conda.

File: mti_nma/steps/nma/nma.py (hunk 8 above)
Comment (jcass11) [reply mti_nma/steps/nma/nma.py]: Hm - then how am I supposed to use the structure-named (non-generic) upstream tasks so that pull will work?

Comment (JacksonMaxfield) [reply mti_nma/steps/nma/nma.py]: Right, this is why I think `pull` is currently broken. This is set in `run`. Not `init`. The way step works is that when you want to do any operation. You init the class then run it there is no statefulness between commands.

Comment (jcass11) [reply mti_nma/steps/nma/nma.py]: Okay I understand - thanks

**Comment by jcass11** [comment]
> Should be able to `git rm ~` or similar. Sorry about that. When I said I had fixed it I meant I had fixed the problem of it creating more files but should have also mentioned to delete those files as well.
> 
> Thanks for making the changes and nice work!

When I try `git rm ~` I get `fatal: /home/juliec: '/home/juliec' is outside repository` which makes me very nervous about messing around with this....

**Comment by jcass11** [comment]
> > Should be able to `git rm ~` or similar. Sorry about that. When I said I had fixed it I meant I had fixed the problem of it creating more files but should have also mentioned to delete those files as well.
> > Thanks for making the changes and nice work!
> 
> When I try `git rm ~` I get `fatal: /home/juliec: '/home/juliec' is outside repository` which makes me very nervous about messing around with this....

For our records, @JacksonMaxfield let me know that:

> It was rm -Rf \~ to escape the normal behavior and look in the current directory for "~" dir

and pushed a commit to address this.

File: mti_nma/steps/avgshape/avgshape.py [hunk 11]
```diff
@@ -108,6 +106,6 @@ def run(self, sh_df=None, struct="Nuc", **kwargs):
 
         # Save manifest as csv
         self.manifest.to_csv(
-            self.step_local_staging_dir / Path(f"manifest_{struct}.csv"), index=False
+            self.step_local_staging_dir / Path(f"manifest.csv"), index=False
```
Comment (vianamp) [inline mti_nma/steps/avgshape/avgshape.py]: I like this :)

File: setup.py [hunk 12]
```diff
@@ -42,14 +42,13 @@
...
     "jupyterlab",
     "matplotlib",
-    "mpl_toolkits"
 ]
 
 requirements = [
-    "aicsimageio",
-    "aicsshparam",
+    "aicsimageio>=3.1.2",
+    "aicsshparam>=0.0.7",
```
Comment (vianamp) [inline setup.py]: Please update this to >=0.0.9.

**Pull request merged**